php/class-customize-snapshot-manager.php

```diff
@@ -914,4 +915,56 @@ public function get_front_uuid_param() {
 	public function get_customize_uuid_param() {
 		return constant( get_class( $this->post_type ) . '::CUSTOMIZE_UUID_PARAM_NAME' );
 	}
+
+	/**
+	 * Let unauthenticated users see posts published in a changeset.
+	 *
+	 * @param array $allcaps All capabilities.
+	 * @param array $caps    Capabilities.
+	 * @param array $args    Args.
+	 * @return array All capabilities.
+	 */
+	public function filter_user_has_cap( $allcaps, $caps, $args ) {
+		global $current_user;
+
+		if (
+			! $this->current_snapshot_uuid
+			|| 0 !== $current_user->ID
+			|| ! isset( $args[2] )
+			|| 'read_post' !== $args['0']
+		) {
+			return $allcaps;
+		}
+
+		$post = get_post( $args[2] );
+		if ( ! $post ) {
+			return $allcaps;
+		}
+		if ( 'customize-draft' !== $post->post_status && 'auto-draft' !== $post->post_status ) {
+			return $allcaps;
+		}
+
+		// Check if the status of the post is 'published' within the changeset.
+		$changeset_id = $this->post_type->find_post( $this->current_snapshot_uuid );
+		if ( ! $changeset_id ) {
+			return $allcaps;
+		}
+		$data = $this->post_type->get_post_content( get_post( absint( $changeset_id ) ) );
+
+		$is_published = false;
+		$key = 'post[' . $post->post_type . '][' . $post->ID . ']';
+
+		if ( isset( $data[ $key ] ) ) {
+			$changeset_post_values = $data[ $key ]['value'];
+			if ( isset( $changeset_post_values['post_status'] ) ) {
+				$is_published = 'publish' === $changeset_post_values['post_status'];
+			}
+		}
+
+		if ( true === $is_published ) {
+			$allcaps[ $caps[0] ] = true;
+		}
```

Comment:
You might want to change the condition here to be:

```php
$allcaps[ $caps[0] ] = $is_published;
```

This could potentially handle the reverse condition where access to a post could be _revoked_ if if its status is changed from `publish` to `private` and the user is not authorized to view private posts. There's probably more to it than that, but this is what came to mind.